17081 Upgrade to Angular 1.3

myabc opened the pull request:
https://community.openproject.org/work_packages/17081

TeatroIO commented:
I've prepared a stage. <a href='http://feature-17081-angular-13.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Click to open.</a>

0xF013 committed:
Upgrade Angular to 1.3

myabc commented:
@0xF013 rebased again (removed strict DI changes)

ulferts commented:
The cuke failure is not directly related to this PR. The problem, some js error for which I haven't found the source yet, was introduced in a18d0b0

ulferts commented:
@myabc @0xF013 just rebased this again on #2283 to see if this also fixes the cukes on travis. It fixed the problem locally.

ulferts commented:
@myabc @0xF013 rebased again to clean up history.

ulferts commented:
The modal does not seem to work correctly with this update:

![image](https://cloud.githubusercontent.com/assets/617519/5266521/9c84be16-7a4a-11e4-89ae-b7768459828a.png)

myabc committed:
Remove unused angular-scenario dependency

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented:
@ulferts I've checked modals and they seem to be working. If you meant the column context menu, then yes, it appears that it no longer scrolls horizontally with the table contents.

0xF013 commented:
@ulferts @myabc I could not reproduce the bug with it being cut, although I confirm it does not follow the horizontal scrolling. maybe some old code got saved

kill vendor/assets/components/
kill frontend/bower_components
kill app/assets/javascript/bundles
bower install
gulp:webpack

myabc commented on frontend/public/templates/components/flash_message.html:
```diff
@@ -1,4 +1,4 @@
-<div ng-if="message !== undefined"
```
@0xF013 this causes some issues with layout btw.

0xF013 replied:
@myabc that is weird. maybe the style attribute should be removed? In might override the ng-hide class.

myabc replied:
@0xF013 I'd prefer to revert this particular change and find a solution that works with `ng-if`.

0xF013 replied:
@myabc ng-if created a new scope so you could prefix all the vars under ng-if with $parent.

myabc commented:
I was confusing the _column context_ and _column header_ menus. I'm seeing the same regression in the  _column context_ menu as @ulferts.

myabc commented:
@0xF013 so the issue is the code that was added in this commit: https://github.com/finnlabs/angular-context-menu/commit/4842e54f53f25496d48e8e935161755bdf679bea

myabc committed:
Correct version of angular-animate specified

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented:
@0xF013 the [$animate API](https://docs.angularjs.org/api/ngAnimate/service/$animate) changed in 1.3. angular-context-menu should be modified something along the following lines:

``` diff
diff --git a/src/angular-context-menu.js b/src/angular-context-menu.js
index d67634a..080eae7 100644
--- a/src/angular-context-menu.js
+++ b/src/angular-context-menu.js
@@ -190,7 +190,7 @@ angular.module('ng-context-menu', [])

         contextMenuPromise.then(function(element) {
           angular.element(element).trap();
-          $animate.enter(element, document.body, element, function() {
+          $animate.enter(element, document.body, element).then(function() {
             adjustPosition(element, pointerPosition);
           });
         });
@@ -199,12 +199,11 @@ angular.module('ng-context-menu', [])
         pointerOffset = getOffset(targetPosition, pointerPosition);
       }
       function adjustPosition($element, pointerPosition) {
-        var win = angular.element($window);
         var viewport = {
           top : win.scrollTop(),
           left : win.scrollLeft()
         };
-        
+
         viewport.right = viewport.left + win.width();
         viewport.bottom = viewport.top + win.height();
         var bounds = $element.offset();
```

(this isn't perfect – if you pay attention, you see a 'flash' before the menu is repositioned to its correct location. Not sure what we can do to get around this).

hschink commented on frontend/bower.json:
```diff
@@ -8,34 +8,35 @@
     "openproject-ui_components": "opf/openproject-ui_components#with-bower",
-    "angular": "~1.2.14",
-    "angular-animate": "~1.2.14",
+    "angular": "~1.3",
```
When I checked out this branch and ran `bower install` I got the following message:

```
Unable to find a suitable version for angular, please choose one:
    1) angular#~1.2.14 which resolved to 1.2.27 and is required by openproject-ui_components#e9750ed700 
    2) angular#^1.2.13 which resolved to 1.2.27 and is required by angular-modal#cc565b7089 
    3) angular#1.2.27 which resolved to 1.2.27 and is required by angular-animate#1.2.27 
    4) angular#~1.2 which resolved to 1.2.27 and is required by angular-busy#4.0.4 
    5) angular#1.3.6 which resolved to 1.3.6 and is required by angular-animate#1.3.6, angular-animate#1.3.6, angular-mocks#1.3.6 
    6) angular#~1.3 which resolved to 1.3.6 and is required by angular-busy#4.1.2 
    7) angular#>=1.2.0 which resolved to 1.3.6 and is required by angular-ui-select2#0.0.5 
    8) angular#~1.x which resolved to 1.3.6 and is required by angular-ui-date#0.0.3 
    9) angular#>= 1.0.8 which resolved to 1.3.6 and is required by angular-ui-router#0.2.11 
    10) angular#>= 1.2.5 which resolved to 1.3.6 and is required by angular-context-menu#0.1.8Prefix the choice with ! to persist it to bower.json
```

though [v1.3.5](https://github.com/angular/angular.js/releases/tag/v1.3.5) seems still be available.

myabc replied:
@hschink It looks like 1.3.6 just got released. Please try again. #2279 will also help with this.

myabc committed:
Correct version of angular-sanitize specified

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Tighten angular version dependency; Bump to 1.3.6

Signed-off-by: Alex Coles <alex@alexbcoles.com>

0xF013 commented:
@myabc I got that flash only the first time and if the menu was invoked close to the right border of the viewport. I have a solution for flickering, I am setting visiblity: hidden before entering and setting visiblity: visible on the animation callback. This allows for correct position and size calculation without seeing the element. Will make a PR to the context menu repository

0xF013 commented:
@myabc @houndci see https://github.com/finnlabs/angular-context-menu/pull/3

myabc committed:
Fix sorting dialog for apparent filterFilter change

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Use ng-if, fix scopes in activityComment directive

Reverts change from `ng-if` to `ng-hide` in 288a8a74.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

hschink commented on frontend/app/work_packages/controllers/dialogs/sorting.js:
```diff
@@ -52,13 +52,16 @@ module.exports = function(sortingModal, $scope, $filter, QueryService, I18n) {
-        sorted = $filter('orderBy')(filtered, 'label');
+    var filterBy = term ? { label: term } : {},
+      filtered = $filter('filter')(getRemainingAvailableColumnsData(), filterBy),
+      sorted = $filter('orderBy')(filtered, 'label');
```
_wuff wuff_ _bark bark_ :wolf: (you see I'm hound): The indentation is off.

myabc replied:
@hschink should I align these attributes with `var filterBy`?

myabc committed:
Use prop('disabled') in activityComment tests

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
DRY variable declarations in activityComment test

* Fix camel-casing to placate JSHint.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Fix flash message layout adjust for ng-hide/show

See change from `ng-if` to `ng-hide` in 288a8a74.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Re-implement broken flashMessage directive tests

Marked as pending in f99b3619.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Update angular-context-menu to 0.1.9

Includes support for Angular 1.3. See:
https://github.com/finnlabs/angular-context-menu/commit/140737d1f6c66f8d0cbcf75d0189a93133962985

Signed-off-by: Alex Coles <alex@alexbcoles.com>

ulferts commented:
@0xF013 those specs have been unstable for some time now, repeatedly flickering on the CI. If, any only if, it makes sense to replace them now with protractor tests, I would be very happy.

Commit:
Fix query menu item activation

0xF013 commented:
@ulferts @myabc AFAIR they have been flickering because specs had some issues with focused elements. Right now they are failing for a legitimate reason. Apparently my fix breaks trap()'s funcionality

myabc commented:
@0xF013 strange. Unfortunately, I can't reproduce the failures locally.

0xF013 commented:
@myabc when I open a column or row context menu, no context menu item is focused nor the original element is focused after the menu is closed. Fixing it atm

myabc committed:
Update angular-context-menu to 0.2.0

Removes animations. See:
https://github.com/0xF013/angular-context-menu/commit/04eb81825d72d84d9249549410c7853477808607

Signed-off-by: Alex Coles <alex@alexbcoles.com>

0xF013 commented:
:beers:
